### DISCUSSION TITLE
Identify target that received request

### DISCUSSION THREAD

**Discussion by artburkart** [Kong/kong#9497] [Help] [answered]
Is there any way to identify which target handled a request?

Let's say I use the "consistency-hashing" balancer, and I'd like to add a header to my response that tells me which upstream target handled my request, how could I obtain this information? There is a world where I could add a proxy shim in front of my upstream target server, but I'd very much prefer if Kong could introduce this information somehow.

I'm not really sure what the ideal place would be, but it seems to me that [`kong.ctx.shared`](https://docs.konghq.com/gateway/latest/plugin-development/pdk/kong.ctx/#main) would be appropriate. I'd simply set the value myself in a pre-function plugin, but I think I'd need access to the balancer to identify the intended target. From what I can tell, there's currently no way to access the balancer or its helper functions.

Thanks!

**Comment by samugi** [comment, marked as answer]
Hello @artburkart,
all the balancer data is stored in the context, at `ngx.ctx.balancer_data` during the balancer phase. After that it is possible, for example, to do:
```
function MyPlugin:header_filter(conf)
  local balancer_data = ngx.ctx.balancer_data

  local header_value = fmt("%s:%s IP: %s target_id: %s",
                           balancer_data.hostname,
                           balancer_data.port,
                           balancer_data.ip,
                           balancer_data.balancer_handle.address.target.id)

  kong.response.set_header("X_BALANCER_DATA", header_value)
end
``` 
I hope that helps.

**Reply by artburkart** [reply]
Oh wow, I'll give that a go this week and follow up. I'd like to mark your answer as the answer, so if I don't follow up, please feel free to ping me again. 🙌

**Reply by artburkart** [reply]
It turns out my mistake was attempting to access this balancer information from within the incorrect phase handler.

This was the code I had originally written:

```diff
apiVersion: configuration.konghq.com/v1
kind: KongPlugin
metadata:
  name: receiving-target
plugin: post-function
config:
-  access:
+ header_filter:
  - |
    kong.log.inspect.on()
    kong.log.inspect(ngx.ctx.balancer_data.balancer_handle)
    kong.log.inspect.off()
```

[Wish I had found this sooner!](https://docs.konghq.com/gateway/latest/plugin-development/custom-logic/#available-contexts) 😅 

Thank you for your help!

**Reply by samugi** [reply]
awesome, @artburkart! Glad it worked!
